### PR TITLE
added src and test folder

### PR DESCRIPTION
**Pull request by snehil1930**

### CODE REVIEW

File: src/Assignment_1/.gitignore [hunk 1]
```diff
@@ -0,0 +1,7 @@
+/Constants.class
```
Comment (amangoyal22) [inline src/Assignment_1/.gitignore]: why are these file in git ignore ?

File: src/Assignment_1/Constants.java [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package Assignment_1;
+
+public class Constants{
```
Comment (amangoyal22) [inline src/Assignment_1/Constants.java:8]: comments missing

File: src/Assignment_1/Constants.java [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+	public final static String quantity = "quantity";
+    public final static String type = "type";
+    public final static String name = "name";
```
Comment (amangoyal22) [inline src/Assignment_1/Constants.java]: static comes before final.

File: src/Assignment_1/ItemCheck.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
+package Assignment_1;
+
+import java.util.*;
+public class ItemCheck {
+	ItemCheck(){
```
Comment (amangoyal22) [inline src/Assignment_1/ItemCheck.java:14]: Learn about pmd and checkstyle

File: src/Assignment_1/ItemCheck.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
+package Assignment_1;
+
+import java.util.*;
+public class ItemCheck {
```
Comment (amangoyal22) [inline src/Assignment_1/ItemCheck.java:7]: Function and class level comments are missing

File: src/Assignment_1/ItemCheck.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+//		Exiting from inputs
+		
+		System.out.println("Thank you purchase!Hope to see you soon..");
+		System.exit(1);
+	}
+	
+	
+	
+	
+	void initials(String[]args)
```
Comment (amangoyal22) [inline src/Assignment_1/ItemCheck.java]: Format of comments
```
  /**
   * <About function>
   *
   * @param Args for the string array for input arguments <define each parameter, define only if taking input>
   * @return <define only if returning>
   */
```

File: src/Assignment_1/ItemCheck.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+		
+//		Constructor of class
+	}
+	
+//	Taking other inputs after pressing Y
+	
+	void nextInputs() throws Exception
+	{
+		System.out.println("Do you want to enter more choice.Enter Y/N");
+		String ch;
```
Comment (amangoyal22) [inline src/Assignment_1/ItemCheck.java]: why declaring ch here if not using

File: src/Assignment_1/ItemCheck.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+		final Scanner sc=new Scanner(System.in);
+		ch=sc.nextLine();
+		while(ch.equalsIgnoreCase("n")==false)
+		{
+			String inputs=sc.nextLine();
```
Comment (amangoyal22) [inline src/Assignment_1/ItemCheck.java]: final missing

File: src/Assignment_1/ItemCheck.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+			String ar[]=inputs.split(" ");
```
Comment (amangoyal22) [inline src/Assignment_1/ItemCheck.java]: final missing

File: src/Assignment_1/ItemCheck.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+			
+//			Inputs taken are verified in Validator
+			
+			item it=new Validator().getDetails(ar);
```
Comment (amangoyal22) [inline src/Assignment_1/ItemCheck.java]: final missing

File: src/Assignment_1/ItemCheck.java [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+	}
+	
+	
+	
+	
+	void initials(String[]args)
+	{
+    	try
+		{
+			item it=new Validator().getDetails(args);
```
Comment (amangoyal22) [inline src/Assignment_1/ItemCheck.java]: final missing

Comment (amangoyal22) [reply src/Assignment_1/ItemCheck.java]: could new Validator be static ? and declared at class level ?

File: src/Assignment_1/ItemCheck.java [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+			it.display();
+			nextInputs();
+		}
+		catch(Exception e)
+		{
+			System.out.println("Invalid Input"+e.getMessage());
+			System.exit(0);
```
Comment (amangoyal22) [inline src/Assignment_1/ItemCheck.java:60]: why exiting system at multiple points ?

File: src/Assignment_1/Main.java [hunk 13]
```diff
@@ -0,0 +1,12 @@
+package Assignment_1;
+
+
+public class Main {
+//	Journey of code starts from here
```
Comment (amangoyal22) [inline src/Assignment_1/Main.java]: comment format?

File: src/Assignment_1/Main.java [hunk 14]
```diff
@@ -0,0 +1,12 @@
+package Assignment_1;
+
+
```
Comment (amangoyal22) [inline src/Assignment_1/Main.java:6]: class level comment and description

File: src/Assignment_1/Main.java [hunk 15]
```diff
@@ -0,0 +1,12 @@
...
+public class Main {
+//	Journey of code starts from here
+//  input Taken from Command line
+	public static void main(String[]args)
+	{
+		ItemCheck ic=new ItemCheck();
```
Comment (amangoyal22) [inline src/Assignment_1/Main.java:14]: final ?

File: src/Assignment_1/Utils.java [hunk 16]
```diff
@@ -0,0 +1,41 @@
+package Assignment_1;
+
+public class Utils {
```
Comment (amangoyal22) [inline src/Assignment_1/Utils.java]: comments

File: src/Assignment_1/Utils.java [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+	private float rawTax(float price)
+	{
+		float tax_of_raw= (float) ((12.5/100.0)*price);
+		System.out.println(tax_of_raw);
+		return tax_of_raw;
+	}
+	
+	private float manufacturedTax(float price)
+	{
+		float rem_tax=(float) ((12.5/100.0)*price);
```
Comment (amangoyal22) [inline src/Assignment_1/Utils.java]: 100 used multiple times. Shouldn't it be a static constant

File: src/Assignment_1/Utils.java [hunk 18]
```diff
@@ -0,0 +1,41 @@
+package Assignment_1;
+
+public class Utils {
+	
+	private float rawTax(float price)
+	{
+		float tax_of_raw= (float) ((12.5/100.0)*price);
```
Comment (amangoyal22) [inline src/Assignment_1/Utils.java]: final missing

File: src/Assignment_1/Utils.java [hunk 19]
```diff
@@ -0,0 +1,41 @@
+package Assignment_1;
+
+public class Utils {
+	
+	private float rawTax(float price)
```
Comment (amangoyal22) [inline src/Assignment_1/Utils.java]: final

File: src/Assignment_1/Utils.java [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+	{
+		float tax_of_raw= (float) ((12.5/100.0)*price);
+		System.out.println(tax_of_raw);
+		return tax_of_raw;
+	}
+	
+	private float manufacturedTax(float price)
```
Comment (amangoyal22) [inline src/Assignment_1/Utils.java]: final

File: src/Assignment_1/Utils.java [hunk 21]
```diff
@@ -0,0 +1,41 @@
...
+	{
+		float rem_tax=(float) ((12.5/100.0)*price);
+		float tax_of_manf=(float) (rem_tax+(2.0/100)*price);
+		return tax_of_manf;
+	}
+	
+	private float importedTax(float price)
+	{
+		float tax_of_import = (float) ((10.0 / 100.0)* price);
```
Comment (amangoyal22) [inline src/Assignment_1/Utils.java]: final

File: src/Assignment_1/Utils.java [hunk 22]
```diff
@@ -0,0 +1,41 @@
...
+	}
+	
+	private float manufacturedTax(float price)
+	{
+		float rem_tax=(float) ((12.5/100.0)*price);
+		float tax_of_manf=(float) (rem_tax+(2.0/100)*price);
+		return tax_of_manf;
+	}
+	
+	private float importedTax(float price)
```
Comment (amangoyal22) [inline src/Assignment_1/Utils.java]: final

File: src/Assignment_1/Utils.java [hunk 23]
```diff
@@ -0,0 +1,41 @@
...
+	{
+		float tax_of_import = (float) ((10.0 / 100.0)* price);
+        if (tax_of_import <= 100) tax_of_import+= 5;
+        else if (tax_of_import <= 200) tax_of_import += 10;
+        else tax_of_import += (5.0 / 100.0) * (tax_of_import + price);
+        return tax_of_import;
+	}
+	
+//	Logic for tax calculations are done
+	public float calculateTax(String type,float price)
```
Comment (amangoyal22) [inline src/Assignment_1/Utils.java]: final

File: src/Assignment_1/Utils.java [hunk 24]
```diff
@@ -0,0 +1,41 @@
...
+	{
+		float tax=0;
```
Comment (amangoyal22) [inline src/Assignment_1/Utils.java]: final

File: src/Assignment_1/Utils.java [hunk 25]
```diff
@@ -0,0 +1,41 @@
...
+		if(type.equalsIgnoreCase(Constants.raw))
```
Comment (amangoyal22) [inline src/Assignment_1/Utils.java:64]: brackets

File: src/Assignment_1/Utils.java [hunk 26]
```diff
@@ -0,0 +1,41 @@
...
+			tax=rawTax(price);
+		else if(type.equalsIgnoreCase(Constants.manufactured)) 
+			tax=manufacturedTax(price);
+		else if(type.equalsIgnoreCase(Constants.imported))
```
Comment (amangoyal22) [inline src/Assignment_1/Utils.java:72]: why declaring a variable can a direct return statement

File: src/Assignment_1/Validator.java [hunk 27]
```diff
@@ -0,0 +1,91 @@
...
+
+import java.util.*;
+public class Validator {
+	
+//	To check whether name is first input or not
+//  and name and type are compulsory inputs
+	
+	private void credCheck(ArrayList<String> args) throws Exception
+	{
+		String name="-"+Constants.name;
```
Comment (amangoyal22) [inline src/Assignment_1/Validator.java]: constant and use concat method for string

Comment (amangoyal22) [reply src/Assignment_1/Validator.java]: final ?

File: src/Assignment_1/Validator.java [hunk 28]
```diff
@@ -0,0 +1,91 @@
...
+		String type="-"+Constants.type;
```
Comment (amangoyal22) [inline src/Assignment_1/Validator.java]: constant and use concat method for string

final ?

File: src/Assignment_1/Validator.java [hunk 29]
```diff
@@ -0,0 +1,91 @@
...
+		if(args.indexOf(name)!=0)
```
Comment (amangoyal22) [inline src/Assignment_1/Validator.java:28]: brackets

File: src/Assignment_1/Validator.java [hunk 30]
```diff
@@ -0,0 +1,91 @@
...
+			throw new Exception("Input should start with -name");
+		if(args.indexOf(name)<0)
+			throw new Exception("Input should have type");
```
Comment (amangoyal22) [inline src/Assignment_1/Validator.java:34]: Give user an example of correct case

File: src/Assignment_1/Validator.java [hunk 31]
```diff
@@ -0,0 +1,91 @@
+package Assignment_1;
+
+import java.util.*;
+public class Validator {
```
Comment (amangoyal22) [inline src/Assignment_1/Validator.java:10]: comments

File: src/Assignment_1/Validator.java [hunk 32]
```diff
@@ -0,0 +1,91 @@
...
+		String type="-"+Constants.type;
+		if(args.indexOf(name)!=0)
+			throw new Exception("Input should start with -name");
+		if(args.indexOf(name)<0)
+			throw new Exception("Input should have type");
+	}
+
+//	Check if input provided for price and quantity are in right format and under the constraints
+	
+	private  void parsingCheck(ArrayList<String> args) throws Exception {
```
Comment (amangoyal22) [inline src/Assignment_1/Validator.java]: why are u throwing exception at function level ?

Comment (amangoyal22) [reply src/Assignment_1/Validator.java]: final missing

File: src/Assignment_1/Validator.java [hunk 33]
```diff
@@ -0,0 +1,91 @@
...
+		String quantity="-"+Constants.quantity;
+		String price="-"+Constants.price;
+		int quant=0;
+		double pr=0;
+        try {
+            if (args.indexOf(quantity) >= 0)
```
Comment (amangoyal22) [inline src/Assignment_1/Validator.java:49]: brackets, null checks

File: src/Assignment_1/Validator.java (hunk 30 above)
Comment (amangoyal22) [reply src/Assignment_1/Validator.java:34]: null checks

File: src/Assignment_1/Validator.java [hunk 34]
```diff
@@ -0,0 +1,91 @@
...
+ 
+        } catch (NumberFormatException e) {
+            System.out.println("Invalid input for price and quantity !");
+        }
+    }
+	
+// Null value check for Type
+	public void nullCheck(Map<String,String> mp) throws Exception
+	{
+		if(mp.get("-"+Constants.type)==null)
```
Comment (amangoyal22) [inline src/Assignment_1/Validator.java]: formatting

Comment (amangoyal22) [reply src/Assignment_1/Validator.java]: Objects method should be used

File: test/Assig1/valdatortester.java [hunk 35]
```diff
@@ -0,0 +1,86 @@
...
+import Assignment_1.*;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+import org.junit.jupiter.api.Test;
+
+class valdatortester {
+
+	@Test
+	void test() throws Exception {
```
Comment (amangoyal22) [inline test/Assig1/valdatortester.java]: make multiple test cases not single test case, with comments and proper naming

File: test/Assig1/valdatortester.java [hunk 36]
```diff
@@ -0,0 +1,86 @@
+package Assig1;
+
+import Assignment_1.*;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+import org.junit.jupiter.api.Test;
+
+class valdatortester {
```
Comment (amangoyal22) [inline test/Assig1/valdatortester.java:13]: comments

File: src/Assignment_1/item.java [hunk 37]
```diff
@@ -0,0 +1,77 @@
...
+	private  String name;
+	private  float price;
+	private  String type;
+	private  int quantity;
+	private float Tax;
+	private float finalPrice;
+	
+//	Constructor
+	
+	public item(final String name,final float price,final int quantity,final String type)
```
Comment (amangoyal22) [inline src/Assignment_1/item.java:26]: format

File: src/Assignment_1/item.java [hunk 38]
```diff
@@ -0,0 +1,77 @@
+package Assignment_1;
+
+import java.util.*;
+public class item {
```
Comment (amangoyal22) [inline src/Assignment_1/item.java:10]: comments

File: src/Assignment_1/item.java [hunk 39]
```diff
@@ -0,0 +1,77 @@
...
+		this.price=price;
+		this.quantity=quantity;
+		this.type=type;
+	}
+	
+	//	summing total price
+	
+	public void totaling()
+	{
+	    Tax=new Utils().calculateTax(type,price);
```
Comment (amangoyal22) [inline src/Assignment_1/item.java:41]: final ?

File: src/Assignment_1/Utils.java (hunk 16 above)
Comment (amangoyal22) [reply src/Assignment_1/Utils.java]: class should be final

**Review by amangoyal22 — CHANGES_REQUESTED** [review]

File: src/Assignment_1/Constants.java [hunk 40]
```diff
@@ -1,11 +1,17 @@
...
 public class Constants{
-	public final static String quantity = "quantity";
-    public final static String type = "type";
-    public final static String name = "name";
-    public final static String price = "price";
-    public final static String raw = "raw";
-    public final static String imported = "imported";
-    public final static String manufactured = "manufactured";
+	public static final String quantity = "quantity";
+    public static final String type = "type";
```
Comment (amangoyal22) [inline src/Assignment_1/Constants.java:13]: Comments for variables ?

File: src/Assignment_1/ItemCheck.java [hunk 41]
```diff
@@ -1,48 +1,55 @@
...
+      */
 	
 	void nextInputs() throws Exception
 	{
 		System.out.println("Do you want to enter more choice.Enter Y/N");
-		String ch;
 		final Scanner sc=new Scanner(System.in);
-		ch=sc.nextLine();
+		String ch=sc.nextLine();
 		while(ch.equalsIgnoreCase("n")==false)
```
Comment (amangoyal22) [inline src/Assignment_1/ItemCheck.java]: constant value comes first

Comment (amangoyal22) [reply src/Assignment_1/ItemCheck.java]: null pointer

File: src/Assignment_1/Utils.java [hunk 42]
```diff
@@ -1,41 +1,66 @@
...
 	
-	private float importedTax(float price)
+	private float importedTax(final float price)
 	{
-		float tax_of_import = (float) ((10.0 / 100.0)* price);
-        if (tax_of_import <= 100) tax_of_import+= 5;
-        else if (tax_of_import <= 200) tax_of_import += 10;
-        else tax_of_import += (5.0 / 100.0) * (tax_of_import + price);
+		float tax_of_import = (float) ((10.0 / Constants.hun)* price);
+        if (tax_of_import <= 100)
```
Comment (amangoyal22) [inline src/Assignment_1/Utils.java]: brackets ?

**Review by amangoyal22 — COMMENTED** [review]
Fix all comments

File: src/Assignment_1/.gitignore (hunk 1 above)
Comment (snehil1930) [reply src/Assignment_1/.gitignore]: removed gitignore file

File: src/Assignment_1/Constants.java (hunk 40 above)
Comment (snehil1930) [reply src/Assignment_1/Constants.java:13]: Done with comments

Comment (amangoyal22) [reply src/Assignment_1/Constants.java:13]: Do formatting right from next time in case of constants

File: src/Assignment_1/ItemCheck.java (hunk 5 above)
Comment (amangoyal22) [reply src/Assignment_1/ItemCheck.java:7]: formatting

File: src/Assignment_1/Validator.java (hunk 31 above)
Comment (amangoyal22) [reply src/Assignment_1/Validator.java:10]: formatting

File: src/Assignment_1/Utils.java (hunk 25 above)
Comment (amangoyal22) [reply src/Assignment_1/Utils.java:64]: formatting

File: src/Assignment_1/Validator.java (hunk 29 above)
Comment (amangoyal22) [reply src/Assignment_1/Validator.java:28]: formatting

File: src/Assignment_1/Validator.java [hunk 43]
```diff
@@ -64,11 +81,13 @@ public void nullCheck(Map<String,String> mp) throws Exception
 	* @param mp hash of arguments
     */
 	
-	public void borderCheck(Map<String,String> mp) throws Exception
+	public void borderCheck(final Map<String,String> mp) throws Exception
 	{
-		String s="-".concat(Constants.type);
-		if (mp.get(s).equals(Constants.raw)==false && mp.get(s).equals(Constants.manufactured)==false && mp.get(s).equals(Constants.imported)==false)
+		final String s=(Constants.dash).concat(Constants.type);
+		if (!mp.get(s).equals(Constants.raw) && !mp.get(s).equals(Constants.manufactured) && !mp.get(s).equals(Constants.imported))
```
Comment (amangoyal22) [inline src/Assignment_1/Validator.java:87]: formatting

File: test/Assig1/valdatortester.java [hunk 44]
```diff
@@ -25,66 +28,94 @@ void validateTest() throws Exception {
...
-        args = new String[]{"-name" , "snehil" , "-type" , "raw" , "-price" , "123" , "-quantity" , "2"} ;
-        vc=new Validator();
+	}
+
+	/*
+	*This method is use to test inputs on various invalid input and test exceptions
+	 */
+
+	@Test
+	void exceptionsTest() throw Exception
```
Comment (amangoyal22) [inline test/Assig1/valdatortester.java:38]: Naming

File: test/Assig1/valdatortester.java (hunk 36 above)
Comment (amangoyal22) [reply test/Assig1/valdatortester.java:13]: format

**Review by amangoyal22 — APPROVED** [review]

**Pull request merged**